File: my-folder/0022-generate-parentheses/solution.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        def dfs(op,cp, stack):
            if ( op == cp == n):
                res.append("".join(stack))
                return
            if ( op < n ):
                stack.append('(')
                dfs( op + 1, cp, stack)
                stack.pop()
            if ( cp < op):
                stack.append(')')
                dfs(op, cp + 1, stack)
                stack.pop()
        dfs(0,0, [])
        return res

            



        

File: my-folder/0022-generate-parentheses/solution.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();
        int op = 0, cp = 0;
        getAllParanthesis(0, 0, n, "", result);
        return result;
    }

    private void getAllParanthesis(int op, int cp, int n, String curr, List<String> result){
        if(curr.length() == n * 2){
            result.add(curr);
            return;
        }

        if(op < n) getAllParanthesis(op + 1, cp, n, curr + '(', result);
        if(cp < op) getAllParanthesis(op, cp + 1, n, curr + ')', result);

    }
}
